File: android/core/protocol/src/main/kotlin/fr/mina/gateway/protocol/ChatCrypto.kt
package fr.mina.gateway.protocol

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.security.SecureRandom
import java.security.Signature
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Chiffrement bout-en-bout du chat — miroir exact de `src/devices/chat-crypto.mjs`.
 *
 * Une clé AES-256 par ÉPOQUE. L'AAD lie chaque ciphertext à son contexte (expéditeur, fil,
 * époque, dates) : déplacer un message chiffré vers un autre en-tête casse le déchiffrement
 * au lieu de passer inaperçu. Une révocation crée une nouvelle époque — l'appareil retiré ne
 * lit plus les événements FUTURS (on ne prétend pas effacer ce qu'il a déjà lu).
 */
object ChatCrypto {
    private const val TRANSFORMATION = "AES/GCM/NoPadding"
    private const val TAG_BITS = 128
    private const val NONCE_BYTES = 12
    private const val KEY_BYTES = 32
    private const val EPOCH_WRAP_PREFIX = "MINA_EPOCH_WRAP_V1"
    private const val ATTACHMENT_INFO = "mina-chat-attachment-v1"
    private val secureRandom = SecureRandom()

    private fun requireKey(key: ByteArray, label: String): ByteArray {
        require(key.size == KEY_BYTES) { "${label}_doit_faire_32_octets" }
        return key
    }

    /** Clé de pièce jointe dérivée de l'époque : une pièce compromise n'expose pas les autres. */
    fun deriveAttachmentKey(epochKey: ByteArray, attachmentId: String): ByteArray =
        hkdfSha256(
            ikm = requireKey(epochKey, "epoch_key"),
            salt = attachmentId.toByteArray(StandardCharsets.UTF_8),
            info = ATTACHMENT_INFO.toByteArray(StandardCharsets.UTF_8),
            length = KEY_BYTES,
        )

    /** HKDF-SHA256 (RFC 5869) — identique à `hkdfSync` de Node. */
    private fun hkdfSha256(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        val extractMac = Mac.getInstance("HmacSHA256")
        extractMac.init(SecretKeySpec(if (salt.isEmpty()) ByteArray(32) else salt, "HmacSHA256"))
        val prk = extractMac.doFinal(ikm)

        val expandMac = Mac.getInstance("HmacSHA256")
        expandMac.init(SecretKeySpec(prk, "HmacSHA256"))
        val out = ByteArrayOutputStream()
        var previous = ByteArray(0)
        var counter = 1
        while (out.size() < length) {
            expandMac.reset()
            expandMac.update(previous)
            expandMac.update(info)
            expandMac.update(counter.toByte())
            previous = expandMac.doFinal()
            out.write(previous)
            counter += 1
        }
        return out.toByteArray().copyOf(length)
    }

    private fun epochWrapAad(deviceId: String, keyEpoch: Int): ByteArray {
        val device = deviceId.toByteArray(StandardCharsets.UTF_8)
        val header = ByteBuffer.allocate(8).putInt(device.size).putInt(keyEpoch).array()
        return EPOCH_WRAP_PREFIX.toByteArray(StandardCharsets.US_ASCII) + byteArrayOf(0) + header + device
    }

    data class WrappedEpochKey(
        val keyEpoch: Int,
        val nonce: String,
        val ciphertext: String,
        val authTag: String,
    )

    fun wrapEpochKey(deviceWrapKey: ByteArray, epochKey: ByteArray, deviceId: String, keyEpoch: Int): WrappedEpochKey {
        requireKey(deviceWrapKey, "device_wrap_key")
        requireKey(epochKey, "epoch_key")
        val nonce = ByteArray(NONCE_BYTES).also { secureRandom.nextBytes(it) }
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(deviceWrapKey, "AES"), GCMParameterSpec(TAG_BITS, nonce))
        cipher.updateAAD(epochWrapAad(deviceId, keyEpoch))
        val sealed = cipher.doFinal(epochKey)
        // GCM concatène le tag à la fin : on le sépare pour rester au format du contrat Node.
        val ciphertext = sealed.copyOfRange(0, sealed.size - 16)
        val tag = sealed.copyOfRange(sealed.size - 16, sealed.size)
        val encoder = Base64.getEncoder()
        return WrappedEpochKey(
            keyEpoch = keyEpoch,
            nonce = encoder.encodeToString(nonce),
            ciphertext = encoder.encodeToString(ciphertext),
            authTag = encoder.encodeToString(tag),
        )
    }

    fun unwrapEpochKey(deviceWrapKey: ByteArray, wrapped: WrappedEpochKey, deviceId: String, keyEpoch: Int): ByteArray {
        requireKey(deviceWrapKey, "device_wrap_key")
        val decoder = Base64.getDecoder()
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(
            Cipher.DECRYPT_MODE,
            SecretKeySpec(deviceWrapKey, "AES"),
            GCMParameterSpec(TAG_BITS, decoder.decode(wrapped.nonce)),
        )
        cipher.updateAAD(epochWrapAad(deviceId, keyEpoch))
        return cipher.doFinal(decoder.decode(wrapped.ciphertext) + decoder.decode(wrapped.authTag))
    }

    /**
     * Vérifie la signature PUIS déchiffre — jamais l'inverse : déchiffrer un contenu non
     * authentifié reviendrait à traiter des octets d'origine inconnue.
     */
    fun verifyAndDecrypt(event: ChatEvent, epochKey: ByteArray, verifyPublicKey: java.security.PublicKey): String {
        requireKey(epochKey, "epoch_key")
        val decoder = Base64.getDecoder()

        val verifier = Signature.getInstance("SHA256withECDSA")
        verifier.initVerify(verifyPublicKey)
        verifier.update(ChatBinaryCodec.encodeSignatureInput(event.copy(signature = "")))
        require(verifier.verify(decoder.decode(event.signature))) { "chat_signature_invalide" }

        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(
            Cipher.DECRYPT_MODE,
            SecretKeySpec(epochKey, "AES"),
            GCMParameterSpec(TAG_BITS, decoder.decode(event.nonce)),
        )
        cipher.updateAAD(ChatBinaryCodec.encodeHeader(event))
        return try {
            String(
                cipher.doFinal(decoder.decode(event.payloadCiphertext) + decoder.decode(event.authTag)),
                StandardCharsets.UTF_8,
            )
        } catch (error: javax.crypto.AEADBadTagException) {
            // Tag invalide OU contexte modifié : dans les deux cas, contenu non fiable.
            throw IllegalStateException("chat_dechiffrement_impossible", error)
        }
    }

    fun randomKey(): ByteArray = ByteArray(KEY_BYTES).also { secureRandom.nextBytes(it) }
}

File: src/devices/chat-crypto.mjs
// Chiffrement bout-en-bout du chat natif (Task 3).
//
// Modèle de clés :
//   - une clé de conversation AES-256 par ÉPOQUE (`keyEpoch`) ; le PC la génère et la tourne ;
//   - chaque appareil possède une `deviceWrapKey` distincte qui sert UNIQUEMENT à recevoir la
//     clé d'époque enveloppée. Une révocation crée une nouvelle époque : l'appareil retiré ne
//     peut plus lire les événements FUTURS (on ne prétend pas effacer ce qu'il a déjà lu).
//
// L'AAD lie chaque ciphertext à son contexte exact (expéditeur, fil, époque, dates) : déplacer
// un message chiffré vers un autre en-tête casse le déchiffrement au lieu de passer inaperçu.

import { createCipheriv, createDecipheriv, createSign, createVerify, hkdfSync, randomBytes } from 'node:crypto';
import { encodeChatHeader, encodeChatSignatureInput } from '../contracts/chat-binary-codec.mjs';

const ALGORITHM = 'aes-256-gcm';
const NONCE_BYTES = 12;
const TAG_BYTES = 16;
const KEY_BYTES = 32;

const EPOCH_WRAP_PREFIX = 'MINA_EPOCH_WRAP_V1';
const ATTACHMENT_INFO = 'mina-chat-attachment-v1';

const requireKey = (value, label) => {
  const key = Buffer.from(value ?? []);
  if (key.length !== KEY_BYTES) throw new TypeError(`${label}_doit_faire_32_octets`);
  return key;
};

/** Clé de pièce jointe dérivée de l'époque : une pièce compromise n'expose pas les autres. */
export const deriveAttachmentKey = ({ epochKey, attachmentId }) => Buffer.from(hkdfSync(
  'sha256',
  requireKey(epochKey, 'epoch_key'),
  Buffer.from(String(attachmentId), 'utf8'),
  Buffer.from(ATTACHMENT_INFO, 'utf8'),
  KEY_BYTES,
));

// AAD binaire domain-separated pour l'enveloppement d'une clé d'époque : la clé enveloppée
// n'est déchiffrable que pour CE couple (appareil, époque).
function epochWrapAad({ deviceId, keyEpoch }) {
  const device = Buffer.from(String(deviceId), 'utf8');
  const header = Buffer.allocUnsafe(4 + 4);
  header.writeUInt32BE(device.length, 0);
  header.writeUInt32BE(Number(keyEpoch), 4);
  return Buffer.concat([Buffer.from(EPOCH_WRAP_PREFIX, 'ascii'), Buffer.from([0]), header, device]);
}

export function wrapEpochKey({ deviceWrapKey, epochKey, deviceId, keyEpoch }) {
  const wrapKey = requireKey(deviceWrapKey, 'device_wrap_key');
  const key = requireKey(epochKey, 'epoch_key');
  const nonce = randomBytes(NONCE_BYTES);
  const cipher = createCipheriv(ALGORITHM, wrapKey, nonce);
  cipher.setAAD(epochWrapAad({ deviceId, keyEpoch }));
  const ciphertext = Buffer.concat([cipher.update(key), cipher.final()]);
  key.fill(0);
  return Object.freeze({
    keyEpoch: Number(keyEpoch),
    nonce: nonce.toString('base64'),
    ciphertext: ciphertext.toString('base64'),
    authTag: cipher.getAuthTag().toString('base64'),
  });
}

export function unwrapEpochKey({ deviceWrapKey, wrapped, deviceId, keyEpoch }) {
  const wrapKey = requireKey(deviceWrapKey, 'device_wrap_key');
  const decipher = createDecipheriv(ALGORITHM, wrapKey, Buffer.from(wrapped.nonce, 'base64'));
  decipher.setAAD(epochWrapAad({ deviceId, keyEpoch }));
  decipher.setAuthTag(Buffer.from(wrapped.authTag, 'base64'));
  return Buffer.concat([decipher.update(Buffer.from(wrapped.ciphertext, 'base64')), decipher.final()]);
}

export function createChatCrypto({ signingPrivateKey, verifyPublicKey, epochKey } = {}) {
  const key = requireKey(epochKey, 'epoch_key');

  return Object.freeze({
    /**
     * Chiffre puis signe. L'ordre compte : on signe le CIPHERTEXT, jamais le clair — la
     * signature prouve l'origine sans qu'un vérificateur ait besoin de déchiffrer.
     */
    encryptAndSign({ header, plaintext }) {
      if (!signingPrivateKey) throw new TypeError('signing_private_key_requise');
      const nonce = randomBytes(NONCE_BYTES);
      const cipher = createCipheriv(ALGORITHM, key, nonce);
      cipher.setAAD(encodeChatHeader({ ...header, version: header.version ?? 2 }));
      const ciphertext = Buffer.concat([cipher.update(Buffer.from(plaintext, 'utf8')), cipher.final()]);
      const event = {
        ...header,
        version: header.version ?? 2,
        payloadCiphertext: ciphertext.toString('base64'),
        nonce: nonce.toString('base64'),
        authTag: cipher.getAuthTag().toString('base64'),
      };
      const signature = createSign('sha256')
        .update(encodeChatSignatureInput({ ...event, signature: '' }))
        .sign({ key: signingPrivateKey, dsaEncoding: 'der' });
      return Object.freeze({ ...event, signature: signature.toString('base64') });
    },

    /**
     * Vérifie PUIS déchiffre — jamais l'inverse. Déchiffrer un contenu non authentifié
     * reviendrait à traiter des octets d'origine inconnue.
     */
    verifyAndDecrypt(event) {
      if (!verifyPublicKey) throw new TypeError('verify_public_key_requise');
      const signature = Buffer.from(event.signature, 'base64');
      const valid = createVerify('sha256')
        .update(encodeChatSignatureInput({ ...event, signature: '' }))
        .verify({ key: verifyPublicKey, dsaEncoding: 'der' }, signature);
      if (!valid) throw new Error('chat_signature_invalide');

      const decipher = createDecipheriv(ALGORITHM, key, Buffer.from(event.nonce, 'base64'));
      decipher.setAAD(encodeChatHeader(event));
      decipher.setAuthTag(Buffer.from(event.authTag, 'base64'));
      try {
        const plaintext = Buffer.concat([
          decipher.update(Buffer.from(event.payloadCiphertext, 'base64')),
          decipher.final(),
        ]);
        return plaintext.toString('utf8');
      } catch {
        // Tag invalide OU contexte modifié (AAD) : dans les deux cas, contenu non fiable.
        throw new Error('chat_dechiffrement_impossible');
      }
    },
  });
}

export const CHAT_CRYPTO_CONSTANTS = Object.freeze({
  ALGORITHM, NONCE_BYTES, TAG_BYTES, KEY_BYTES, EPOCH_WRAP_PREFIX, ATTACHMENT_INFO,
});
